torch/csrc/autograd/engine.cpp

```diff
@@ -782,8 +782,7 @@ void set_device(int device) {
     for (const auto i : c10::irange(static_cast<size_t>(
              c10::DeviceType::COMPILE_TIME_MAX_DEVICE_TYPES))) {
       auto* impl = c10::impl::device_guard_impl_registry[i].load();
-      if (impl && device < impl->deviceCount() &&
-          impl->getDevice().index() != device) {
+      if (impl && device < impl->deviceCount()) {
```

Comment:
can we call `.index()` only if `deviceCount() > 1`? That would preserve behavior that we want for cuda, and won't interfere with devices that don't have `index`

Reply:
Hi, thanks for your reply. With the current code, when switch a new thread with the same device, it is unnecessary to initialize to execute  `set_device`. This is good for cuda, but for other devices, especially for devices with insufficient multithreading support, it is a very bad thing, because `setdevice` is required to initialize.
So if you want to  preserve this behavior, can we add the macros `USE_CUDA` to do this? Just like the fllowing:
```
#if defined(USE_CUDA)
      if (impl && device < impl->deviceCount() &&
          impl->getDevice().index() != device) {
        impl->setDevice(at::Device(static_cast<c10::DeviceType>(i), device));
      }
#else
      if (impl && device < impl->deviceCount()) {
        impl->setDevice(at::Device(static_cast<c10::DeviceType>(i), device));
      }
#endif
```

Reply:
`USE_CUDA` is not defined in `engine.cpp`, that's the reason the original PR was so hard to make

Reply:
> `USE_CUDA` is not defined in `engine.cpp`, that's the reason the original PR was so hard to make

Sorry, I haven't noticed this limitation yet. ehh, is that ok to add a check with device.is_cuda() to replace the macro `USE_CUDA`. This will not affect the behavior of cuda, and will be more friendly to other devices,  just as the fllowing:

```
     if (impl && device.is_cuda() && device < impl->deviceCount()  &&  impl->getDevice().index() != device ) {
                impl->setDevice(at::Device(static_cast<c10::DeviceType>(i), device));
     } else {
            impl->setDevice(at::Device(static_cast<c10::DeviceType>(i), device));
     }
```
@ngimel , does it sound good to you ?

Reply:
It's fine to leave as is for now I think. We were trying to avoid `Device` construction when it's not needed, but constructing `Device` is cheap, so it's not a big regression. I still don't understand why calling `.index()` won't work in your case.

Reply:
@ngimel ,sorry, maybe I didn't express clearly. The reason for my modification is not that calling `.index` does not work， it is `impl->getDevice`. There are two reasons for my changes, the first is that the behavior here is inconsistent with previous versions. The second is that for my case (custom device), the multi-thread support is not so perfect. After switching the new thread, it needs to be initialized by `setDevice`, and the operation similar to `getDevice` cannot be performed first.